### PR TITLE
Remove old scheduler and friends

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This PR removes the legacy Ray scheduler. Most of the removal is straightforward, the only significant bit is the part removing the reconstruction policy.
- Previously, the reconstruction policy was responsible for marking objects as failed (tested in test_reference_counting_2.py::test_borrowed_id_failure)
- In this PR, this job is done by the OwnershipBasedObjectDirectory when it notices the owner is dead.

Note that this does mean after this PR we have to roll forward with OwnershipBasedObjectDirectory; the other object directory will no longer be handling failures.

### CODE REVIEW

**Comment by ericl** [comment]
@wuisawesome you can try this as a starting point, though according to @stephanie-wang we need to keep reconstruction policy.

File: src/ray/raylet/raylet.cc [hunk 1]
```diff
@@ -66,8 +66,17 @@ Raylet::Raylet(boost::asio::io_service &main_service, const std::string &socket_
       object_directory_(
           RayConfig::instance().ownership_based_object_directory_enabled()
               ? std::dynamic_pointer_cast<ObjectDirectoryInterface>(
-                    std::make_shared<OwnershipBasedObjectDirectory>(main_service,
-                                                                    gcs_client_))
+                    std::make_shared<OwnershipBasedObjectDirectory>(
+                        main_service, gcs_client_,
+
+                        [this](const ObjectID &obj_id) {
+                          rpc::ObjectReference ref;
```
Comment (stephanie-wang) [inline src/ray/raylet/raylet.cc:73]: TODO to return error to the client instead (putting the error in the object store is a bit sketchy if there's not enough memory).

**Review by stephanie-wang — APPROVED** [review]

File: src/ray/raylet/raylet.cc (hunk 1 above)
Comment (ericl) [reply src/ray/raylet/raylet.cc:73]: Done

File: src/ray/core_worker/store_provider/plasma_store_provider.cc [hunk 2]
```diff
@@ -239,16 +239,12 @@ Status CoreWorkerPlasmaStoreProvider::GetIfLocal(
...
-    // NOTE: for direct call actors, we still need to issue an unblock IPC to release
-    // get subscriptions, even if the worker isn't blocked.
-    if (ctx.ShouldReleaseResourcesOnBlockingCalls() || ctx.CurrentActorIsDirectCall()) {
-      return client->NotifyDirectCallTaskUnblocked();
-    } else {
-      return Status::OK();  // We don't need to release resources.
-    }
+  // NOTE: for direct call actors, we still need to issue an unblock IPC to release
+  // get subscriptions, even if the worker isn't blocked.
+  if (ctx.ShouldReleaseResourcesOnBlockingCalls() || ctx.CurrentActorIsDirectCall()) {
```
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: We still need it, I added a TODO in node_manager that we can remove it after we fix the issue.

**Comment by stephanie-wang** [comment]
Does this also close #14458 then?

**Comment by ericl** [comment]
I think so, though the OBOD will still be calling GetObjectStatus. But at least there isn't polling.

**Pull request merged**